### PR TITLE
Changed functions that iterate and call getNumberDensity many times to more optimal methods 

### PR DESCRIPTION
**Pull request by ZanderUF**
Changed calls to getNumberDensity in loops to instead use  getNumberDensities or getNuclideNumberDensities.  This avoids some overhead in getNumberDensity that does not need to be repeated during iterative processes.

### CODE REVIEW

File: armi/reactor/blocks.py [hunk 1]
```diff
@@ -757,8 +757,11 @@ def adjustDensity(self, frac, adjustList, returnMass=False):
         if returnMass:
             # do this with a flag to enable faster operation when mass is not needed.
             volume = self.getVolume()
-        for nuclideName in adjustList:
-            dens = self.getNumberDensity(nuclideName)
+
+        numDensities = self.getNuclideNumberDensities(adjustList)
+
+        for nuclideName,dens in zip(adjustList,numDensities):
```
Comment (ntouran) [inline armi/reactor/blocks.py:763]: Lack of space after `,` indicates that you should run `black` on this file to get formatting perfect.

File: armi/reactor/blocks.py [hunk 2]
```diff
@@ -1212,15 +1215,17 @@ def mergeWithBlock(self, otherBlock, fraction):
         into a full block.
         """
 
+        numDensities = self.getNumberDensities()
```
Comment (ntouran) [inline armi/reactor/blocks.py:1218]: Going even further, this method and the `self.getNuclides()` call both have to run `getNuclides()` which is redundant. I recommend:

```
nuclideNames = self.getNuclides();
numDensities = self.getNuclideNumberDensities()

for nucName, ndens in zip(nucNames, nucDensities):
   set.setNum...
```

Oohh and if you want to get crazy you can use `setNumberDensities` to speed things up on the setting side. Though really that could be a different diff.

File: armi/reactor/blocks.py [hunk 3]
```diff
@@ -1212,15 +1215,17 @@ def mergeWithBlock(self, otherBlock, fraction):
...
         for nuc in self.getNuclides():
-            self.setNumberDensity(nuc, (1.0 - fraction) * self.getNumberDensity(nuc))
+            self.setNumberDensity(nuc, (1.0 - fraction) * numDensities[nuc])
 
         # now add the other blocks densities.
         for nuc in otherBlock.getNuclides():
             self.setNumberDensity(
                 nuc,
-                self.getNumberDensity(nuc)
+                numDensities[nuc]
```
Comment (ntouran) [inline armi/reactor/blocks.py:1228]: similar zip down here

File: armi/reactor/blocks.py [hunk 4]
```diff
@@ -1462,9 +1467,11 @@ def getMfp(self, gamma=False):
         mfpNumerator = numpy.zeros(len(flux))
         absMfpNumerator = numpy.zeros(len(flux))
         transportNumerator = numpy.zeros(len(flux))
+
+        numDensities = self.getNumberDensities()
         # vol = self.getVolume()
         for nuc in self.getNuclides():
```
Comment (ntouran) [inline armi/reactor/blocks.py:1473]: can do similar single `getNuclides` and then use `getNuclideNumberDensities` as recommended above.

File: armi/reactor/composites.py [hunk 5]
```diff
@@ -2174,9 +2174,12 @@ def getAtomicWeight(self):
         """
         numerator = 0.0
         denominator = 0.0
+
+        numDensities = self.getNumberDensities()
+
         for nucName in self.getNuclides():
             atomicWeight = nuclideBases.byName[nucName].weight
```
Comment (ntouran) [inline armi/reactor/composites.py:2181]: and again!

File: armi/reactor/converters/geometryConverters.py [hunk 6]
```diff
@@ -910,9 +910,12 @@ def createHomogenizedRZTBlock(
                     )
                 homBlockVolume += blockVolumeHere
                 homBlockTemperature += b.getAverageTempInC() * blockVolumeHere
-                for nucName in b.getNuclides():
+
+                numDensities = b.getNuclideNumberDensities(b.getNuclides())
+
+                for nucName, numDen in zip(b.getNuclides(), numDensities):
```
Comment (ntouran) [inline armi/reactor/converters/geometryConverters.py:916]: recommending reuse the `getNuclides` resturn from above

**Review by ntouran — CHANGES_REQUESTED** [review]
Just a few small requests, plus run the code formatter

File: armi/reactor/blocks.py (hunk 1 above)
Comment (ZanderUF) [reply armi/reactor/blocks.py]: Right, forgot about that.  Should be done.

**Review by ntouran — COMMENTED** [review]
wonderful, thank you!!

**Review by ntouran — APPROVED** [review]

**Pull request merged**